README.md
# C2 Norm Optimizer

The Norm Optimizer is a C2 VALAWAI component that computes normative systems
that are optimal with respect to some value of interest, using off-the-shelf
metaheuristics optimizers.

Similarly to the [Alignment
Calculator](https://github.com/VALAWAI/C2_alignment_calculator), the Norm
Optimizer takes as input an arbitrary model (which in practice is implemented as
a class) with a method that defines how a set of norms governs the transition at
a time step.

To understand what the Norm Optimizer actually optimizes, it is recommended to
read first the introduction to the [Alignment
Calculator](https://github.com/VALAWAI/C2_alignment_calculator). The Norm
Optimizer simply optimizes the normative parameters in the normative system $N$
with respect to some value of interest given through its *value semantics
function*. The optimization is subject to the lower and upper bounds of the
normative parameters, as well as any potential constraints that should be taken
into account.

# Summary

 - Type: C2
 - Name: Norm Optimizer
 - Version: 1.0.0 (September 27, 2023)
 - API: [1.0.0 (October 4, 2023)](https://editor-next.swagger.io/?url=https://raw.githubusercontent.com/VALAWAI/C2_norm_optimizer/main/component-api.yml)
 - VALAWAI API: [0.1.0 (September 18, 2023)](https://editor-next.swagger.io/?url=https://raw.githubusercontent.com/VALAWAI/MOV/main/valawai-api.yml)
 - Developed by: IIIA-CSIC
 - License: [MIT](LICENSE)

# Usage

To understand how to set up and use this component, the reader is directed to
the `example.py` script available in this repository. There, you can find first
the `TaxModel` class. Read first the **Usage** section of the [Alignment
Calculator](https://github.com/VALAWAI/C2_alignment_calculator) to understand
what are the classes, methods and function defined in this script and what is
their purpose.

In the `TaxModel`, the simulation is regulated by two norms, `pay` and
`payback`. Each of these norms is attached to 5 normative parameters, `r0`, ...,
`r4`. Hence, the dictionary of normative parameters that we seek to optimize is:

```python
{'pay': ['r0', 'r1', 'r2', 'r3', 'r4'], 'payback': ['r0', 'r1', 'r2', 'r3', 'r4']}
```

For the optimization search, it is necessary to state the bounds of the
normative parameters:

```python
lower_bounds = {
    'pay': {'r0': 0.0, 'r1': 0.0, 'r2': 0.0, 'r3': 0.0, 'r4': 0.0},
    'payback': {'r0': 0.0, 'r1': 0.0, 'r2': 0.0, 'r3': 0.0, 'r4': 0.0}
}
upper__bounds = {
    'pay': {'r0': 1.0, 'r1': 1.0, 'r2': 1.0, 'r3': 1.0, 'r4': 1.0},
    'payback': {'r0': 1.0, 'r1': 1.0, 'r2': 1.0, 'r3': 1.0, 'r4': 1.0}
}
```

Finally, the last necessary step is to define constraints on the normative
parameters, if any. Mathematically, these constraints are inequalities $f_i(N)
\leq 0$, i.e. that take as input a normative system with their normative
parameters. To set up the Norm Optimizer, we define constraints as functions
that take as input a norm dictionary. For example, the following function
captures the constraints that the rates of `payback` should add up to unity:

```python
def payback_constraint(norms):
    sum_rates = sum(norms["payback"].values())
    return abs(sum_rates - 1.)
```

These constraints are added as a penalty on the fitness function used for the
optimization search.

The Alignment Calculator component is implemented as a
[Flask](https://flask.palletsprojects.com/en/2.3.x/) application. To initialize
it, use the `create_app` function:

```python
from app import create_app

app = create_app(
    YourModel,                      # your model class
    [...],                          # your model initialization arguments
    {...},                          # your model initialization keyword arguments
    norms,                          # your norms dictionary
    your_value_semantics_funcion,   # your value semantics function

    lower_bounds,                   # your lower bounds
    upper_bounds,                   # your upper bounds
    const=[payback_constraint],     # your constraints
    term_dict={"max_early_stop": 1, "epsilon": 0.01}, # termination conditions, see the MEALPY docs
    path_length=10,                 # change if needed, default is 10
    path_sample=500                 # change if needed, default is 500
)
```

To develop your own model to be deployed in a component, the `template.py`
script is provided as a blueprint.

This component communicates through the following HTTP requests:

* Data messages:

    - GET `/opt_norms`

* Control messages:

    - PATCH `/opt_cls` changes MEALPY optimizer used for the search
    - PATCH `/opt_args` changes the initialization arguments for the MEALPY
      optimizer
    - PATCH `/opt_kwargs` changes the initialization keyword arguments for the
      MEALPY optimizer
    - PATCH `/term_dict` changes the termination conditions for the optimization
      search
    - PATCH `/path_length` changes the length of the paths used to sample
      outcomes
    - PATH `/path_sample` changes the number of paths sampled to compute the
      alignment

# Deployment

Clone this repository and develop your model and value semantics functions
following the blueprint in `template.py`:

```bash
$ git clone https://github.com/VALAWAI/C2_norm_optimizer.git
```

Build your Docker image in your directory of the component repository:

```bash
$ cd /path/to/C2_norm_optimizer
$ docker build -t c2_norm_optimizer .
```

Run a Docker container with your C2 Norm Optimizer:

```bash
$ docker run --rm -d \
  --network valawai \
  --name c2_norm_optimizer \
  --mount type=bind,src="$(pwd)",target=/app \
  -p 5432:5000 \
  -e MODEL=my_model \
  c2_norm_optimizer
```

docker run --rm -d \
  --name c2_norm_optimizer \
  --mount type=bind,src="$(pwd)",target=/app \
  -p 5432:5000 \
  -e MODEL=example \
  c2_norm_optimizer

The environment variable `MODEL` refers to the script where you have defined
your model (do not include the .py extension).

Once the container is up and running, use `curl` to communicate with the
component:

```bash
$ curl http://localhost:5432/opt_norms
{
  "algn": 0.9048607793661254,
  "norms": {
    "pay": {
      "r0": 0.5477665900888846,
      "r1": 0.9386083634952787,
      "r2": 0.5055107094173359,
      "r3": 0.9499003310249331,
      "r4": 0.6815803866882211
    },
    "payback": {
      "r0": 0.03399436561190139,
      "r1": 0.3205872727529152,
      "r2": 0.21373772788216727,
      "r3": 0.3620575033734875,
      "r4": 0.1426227309170106
    }
  }
}
```

Change the MEALPY optimizer:

```bash
$ curl -X PATCH http://localhost:5432/opt_cls -d mealpy.physics_based.SA.GaussianSA
{}
```

Change the optimizer's initialization arguments:

```bash
$ curl -X PATCH http://localhost:5000/opt_args -H 'Content-Type: application/json' -d '[10000,100,0.95,0.025]'
```

Change the sample of paths used to compute the alignment:

```bash
$ curl -X PATCH http://localhost:5432/path_sample -d 200
{}
```


CHANGELOG.md
# C2_norm_optimizer
 
## Version 1.0.0 (October 4, 2023)

- Initial component version.


example.py
import argparse
from mesa import Agent, Model
from mesa.time import BaseScheduler
from random import seed, uniform

from app import create_app

parser = argparse.ArgumentParser()
parser.add_argument('port', type=int)

seed(100)

class TaxModel(Model):
    def __init__(self) -> None:
        super().__init__()
        self.num_agents = 200
        self.num_segments = 5
        self.schedule = BaseScheduler(self)
        for i in range(self.num_agents):
            a = Agent(i, self)
            a.wealth = uniform(0, 100)
            self.schedule.add(a)
        self.sorted_agents = sorted(self.schedule.agents, key=lambda a: a.wealth)
        self.ags_per_seg = self.num_agents//self.num_segments
        self.segments = [i//self.ags_per_seg for i in range(self.num_agents)]

    def step(self, norms) -> None:
        common_fund = 0.
        for i in range(self.num_agents):
            a = self.sorted_agents[i]
            seg = self.segments[i]
            pay = a.wealth * norms['pay'][f"r{seg}"]
            if pay > a.wealth:
                pay = a.wealth
            a.wealth -= pay
            common_fund += pay
        common_fund *= 1.05
        segs_payback = [
            common_fund * norms['payback'][f"r{seg}"] / self.ags_per_seg
            for seg in range(self.num_segments)
        ]
        for i in range(self.num_agents):
            a = self.sorted_agents[i]
            seg = self.segments[i]
            a.wealth += segs_payback[seg]
        self.sorted_agents = sorted(self.schedule.agents, key=lambda a: a.wealth)


def ratio_wealth_value(mdl: TaxModel) -> float:
    """Semantics function for value equality based on the ratio between the
    maximum and the minimum wealth.

    Parameters
    ----------
    mdl : TaxModel

    Returns
    -------
    float
    """
    a = mdl.sorted_agents[99].wealth
    b = mdl.sorted_agents[199].wealth
    return a/b


def gini_index_value(mdl: TaxModel) -> float:
    """Semantics function for value equality based on the Gini Index.

    Parameters
    ----------
    mdl : TaxModel

    Returns
    -------
    float
    """
    sorted_wealth = [ag.wealth for ag in mdl.sorted_agents]
    N = mdl.num_agents
    B = sum(xi * (N - i) for i, xi in enumerate(sorted_wealth)) 
    GI =  1 + (1 / N) - 2 * B / (N * sum(sorted_wealth))
    return 1 - 2*GI

        
if __name__ == '__main__':

    norms = {
        'pay': [f'r{i}' for i in range(5)],
        'payback': [f'r{i}' for i in range(5)]
    }
    lower_bounds = {
        'pay': {f'r{i}': 0. for i in range(5)},
        'payback': {f'r{i}': 0. for i in range(5)}
    }
    upper_bounds = {
        'pay': {f'r{i}': 1. for i in range(5)},
        'payback': {f'r{i}': 1. for i in range(5)}
    }

    def payback_constraint(norms):
        sum_rates = sum(norms["payback"].values())
        return abs(sum_rates - 1.)

    app = create_app(
        TaxModel,
        [],
        {},
        norms,
        ratio_wealth_value,
        lower_bounds,
        upper_bounds,
        const=[payback_constraint],
        term_dict={'max_epoch': 1}          # DEBUG only one search iteration
        # term_dict={"max_early_stop": 1, "epsilon": 0.01}
    )

    args = parser.parse_args()
    app.run(debug=True, host='0.0.0.0', port=args.port)


template.py
import argparse
from app import create_app

from typing import Any

parser = argparse.ArgumentParser()
parser.add_argument('port', type=int)

class YourModel:
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        """A model whose evolution is determined by a set of norms."""
        super().__init__(*args, **kwargs)

    def step(
        self,
        norms: dict[str, dict[str, Any]]
    ) -> None:
        """The model evolves according to the set of norms in place.

        By having the norms as an input to the model's ``step()`` method, it is
        possible to model changes in the implemented normative system while the
        model is still running.

        Parameters
        ----------
        norms : Dict[Any, Dict[Any, Any]]
            A map of the norms governing the evolution of the model at that
            step. This map is provided as a dictionary from norm identifiers
            (keys) to a dictionary mapping each of the norms' parameters to
            their values.

        Raises
        ------
        NotImplementedError
        """
        raise NotImplementedError
    

def your_value_semantics_funcion(mdl: YourModel) -> float:
    """Value semantics function.

    Parameters
    ----------
    mdl : YourModel

    Returns
    -------
    float
        The degree of respect for the value at the current state of the model.

    Raises
    ------
    NotImplementedError
    """
    raise NotImplementedError


if __name__ == '__main__':

    norms = {
        'n1': ['p11',...]
    }
    lower_bounds = {
        'n1': {'p11': ...}
    }
    upper_bounds = {
        'n1': {'p11': ...}
    }

    def your_constraint(norms):
        raise NotImplementedError

    app = create_app(
        YourModel,                      # your model class
        [...],                          # your model initialization arguments
        {...},                          # your model initialization keyword arguments
        norms,                          # your norms dictionary
        your_value_semantics_funcion,   # your value semantics function
        lower_bounds,                   # lower bounds for normative parameters
        upper_bounds,                   # upper bounds for normative parameters
        const=[your_constraint,...],    # constraints among normative parameters, if any
        # term_dict={"epsilon": 0.01},  # optimizer termination conditions
        # path_length=10,               # change if needed, default is 10
        # path_sample=500               # change if needed, default is 500
    )

    args = parser.parse_args()
    app.run(debug=True, host="0.0.0.0", port=args.port)


app.py
# Copyright 2023 IIIA-CSIC
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License version 3
# along with this program; if not, see <http://www.gnu.org/licenses/>.

from dataclasses import dataclass
from pydoc import locate
from flask import Flask, Request, jsonify, request
from mealpy.optimizer import Optimizer
from mealpy.evolutionary_based.GA import BaseGA
from traceback import format_exc
from valalgn.sampling import optimize_norms

from typing import Any, Callable, Iterable, Type

@dataclass
class NormOptimizer:
    model_cls: Type
    model_args: list[Any]
    model_kwargs: dict[str, Any]
    norms: dict[str, Iterable[str]]
    value: Callable[[object], float]
    lower_bounds: dict[str, dict[str, float]]
    upper_bounds: dict[str, dict[str, float]]
    const: Iterable[Callable[[dict[str, dict[str, float]]], float]]
    lambda_const: float
    opt_cls: type[Optimizer]
    opt_args: list[Any]
    opt_kwargs: dict[str, Any]
    term_dict: dict[str, Any]
    path_length: int
    path_sample: int

    def optimal_norms(self) -> float:
        return optimize_norms(
            self.model_cls,
            self.model_args,
            self.model_kwargs,
            self.norms,
            self.value,
            self.lower_bounds,
            self.upper_bounds,
            const=self.const,
            lambda_const=self.lambda_const,
            opt_cls=self.opt_cls,
            opt_args=self.opt_args,
            opt_kwargs=self.opt_kwargs,
            term_dict=self.term_dict,
            path_length=self.path_length,
            path_sample=self.path_sample
        )

def create_app(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    norms: dict[str, Iterable[str]],
    value: Callable[[object], float],
    lower_bounds: dict[str, dict[str, float]],
    upper_bounds: dict[str, dict[str, float]],
    const: Iterable[Callable[[dict[str, dict[str, float]]], float]] = [],
    lambda_const: float = 1.,
    opt_cls: type[Optimizer] = BaseGA,
    opt_args: list[Any] = [],
    opt_kwargs: dict[str, Any] = {},
    term_dict: dict[str, Any] = {"max_epoch": 100},
    path_length: int = 10,
    path_sample: int = 500,
) -> Flask:
    """Create a Flask app that optimizes norms with respect to a value.

    This C2 component of the VALAWAI architecture optimizes a set of norms with
    respect to a value (or an aggregation of values) [1]_.

    Parameters
    ----------
    model_cls : Type
        The model (e.g. an agent-based model) governed by a set of norms.
    model_args : list[Any]
        Initialization arguments for the model.
    model_kwargs : dict[str, Any]
        Initialization keyword arguments for the model.
    norms : dict[str, Iterable[str]]
        The identifies of the norms and their parameters that determine the
        evolution of the model.
    value : Callable[[object], float]
        The value semantics function with respect to whom the alignment of the
        norms is optimized.
    lower_bounds : dict[str, dict[str, float]]
        The lower bounds for the normative parameters.
    upper_bounds : dict[str, dict[str, float]]
        The upper bounds for the normative parameters.
    const : Iterable[Callable[[dict[str, dict[str, float]]], float]], optional
        Any constraints that should be respeted by the normative parameters, by
        default []. These are a set of function that take as input a norm
        dictionary and output a real number. The closer the output is to 0, the
        more respectful of the constraints is the normative system.
    lambda_const : float, optional
        The penalty to the fitness of violating the constraints, by default 1.
    opt_cls : type[Optimizer], optional
        The mealpy optimizer class used for the search, by default BaseGA.
    opt_args : list, optional
        The mealpy optimizer initialization arguments, by default [].
    opt_kwargs : dict, optional
        The mealpy optimizer initialization keyword arguments, by default [].,
        by default {}.
    term_dict : dict, optional
        The dictionary for termination condition for the optimization search, by
        default {"max_epoch": 100}. For details, see the mealpy page.
    path_length : int, optional
        The length of the evolution path used to compute the alignment, by
        default 10.
    path_sample : int, optional
        The number of paths to sample when computing the alignment, by
        default 500.

    Returns
    -------
    Flask
        A Flask application that can process GET /opt_norms requests.

    References
    ----------
    .. [1] Montes, N., & Sierra, C. (2022). Synthesis and properties of
        optimally value-aligned normative systems. Journal of Artificial
        Intelligence Research, 74, 1739–1774. https://doi.org/10.1613/jair.1.
        13487
    """

    app = Flask(__name__)
    
    optimizer = NormOptimizer(
        model_cls, model_args, model_kwargs, norms, value, lower_bounds,
        upper_bounds, const, lambda_const, opt_cls, opt_args, opt_kwargs,
        term_dict, path_length, path_sample
    )

    def __check_request(request: Request):
        if not request.is_json:
            return {"error": "Request must be JSON"}, 415
        input_data = request.get_json()
        if not isinstance(input_data, dict):
            return {"error": f"Params must be passed as a dict"}, 400
        return input_data

    @app.get('/opt_norms')
    def get_opt_norms():
        try:
            opt_norms, algn = optimizer.optimal_norms()
            return jsonify({'norms': opt_norms, 'algn': algn}), 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/opt_cls')
    def patch_opt_cls():
        input_data = request.get_data()
        opt_cls_str = input_data.decode()
        try:
            optimizer.opt_cls = locate(opt_cls_str)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/opt_args')
    def patch_opt_args():
        try:
            if not request.is_json:
                return {"error": "Request must be JSON"}, 415
            input_data = request.get_json()
            if not isinstance(input_data, list):
                return {"error": f"Params must be passed as a list"}, 400
            optimizer.opt_args = input_data
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/opt_kwargs')
    def patch_opt_kwargs():
        try:
            opt_kwargs = __check_request(request)
            optimizer.opt_kwargs = opt_kwargs
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400

    @app.patch('/term_dict')
    def patch_term_dict():
        try:
            term_dict = __check_request(request)
            optimizer.term_dict = term_dict
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400

    @app.patch('/path_length')
    def patch_path_length():
        input_data = request.get_data()
        try:
            optimizer.path_length = int(input_data)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/path_sample')
    def patch_path_sample():
        input_data = request.get_data()
        try:
            optimizer.path_sample = int(input_data)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    return app
